lib/service/service.go

```diff
@@ -3299,6 +3332,31 @@ func proxySettingsFromConfig(cfg *Config, proxySSHAddr utils.NetAddr) webclient.
 	return proxySettings
 }
 
+// getPostgresPublicAddr returns the proxy PostgresPublicAddrs based if proxy postgres service
+// was configured on separate listener. For backward compatibility if PostgresPublicAddrs was not provided.
+// Proxy will reuse the PostgresPublicAddrs field to propagate postgres service address to legacy tsh clients.
+func getPostgresPublicAddr(cfg *Config) string {
+	if len(cfg.Proxy.PostgresPublicAddrs) > 0 {
+		return cfg.Proxy.PostgresPublicAddrs[0].String()
+	}
+
+	if len(cfg.Proxy.PostgresPublicAddrs) != 0 || cfg.Proxy.PostgresAddr.IsEmpty() {
+		return ""
+	}
+
+	// DELETE IN 9.0.0
+	// If the PostgresPublicAddrs address was not set propagate separate postgres service listener address
+	// to legacy tsh clients reusing PostgresPublicAddrs field.
+	var host string
+	if len(cfg.Proxy.PublicAddrs) > 0 {
+		// Get proxy host address from public address.
+		host = cfg.Proxy.PublicAddrs[0].Host()
+	} else {
+		host = cfg.Proxy.WebAddr.Host()
+	}
+	return fmt.Sprintf("%s:%d", host, cfg.Proxy.PostgresAddr.Port(defaults.PostgresListenPort))
```

Comment:
Is there a way to reuse the port calculation logic from `getPostgresDefaultPort` you have in `lib/config/configuration.go` above? Seems like it does basically the same thing. And that function already operates on `service.Config` which you have here.

So I would move that function here (I don't think `fc` is really needed there - I left a comment there) and use both here and in lib/config as well so we don't end up with duplication.

Maybe even make a helper method on the `service.Config` struct.

Reply:
> Is there a way to reuse the port calculation logic from getPostgresDefaultPort you have in lib/config/configuration.go above?

The`getPostgresDefaultPort` function returns the postgres port. In this function if the PostgresPublicAddrs was not set  the port is constant - taken from `cfg.Proxy.PostgresAddr.Port` thus  I don't see how this logic can be combined.